src/requestManager.cpp

```diff
@@ -394,67 +400,77 @@ void CRequestManager::SendRequests()
+      if (IsChainNearlySyncd())
+          MIN_BLK_REQUEST_RETRY_INTERVAL = 5*1000*1000;
+      else
+          MIN_BLK_REQUEST_RETRY_INTERVAL = 30*1000*1000;
 
       if (now-item.lastRequestTime > MIN_BLK_REQUEST_RETRY_INTERVAL)  // if never requested then lastRequestTime==0 so this will always be true
 	{
           if (!item.availableFrom.empty())
 	    {
 	      CNodeRequestData next;
-              while ((!item.availableFrom.empty())&&(next.node == NULL)) // Go thru the availableFrom list, looking for the first node that isn't disconnected
+              while (!item.availableFrom.empty() && (next.node == NULL)) // Go thru the availableFrom list, looking for the first node that isn't disconnected
                 {
   	        next = item.availableFrom.front();  // Grab the next location where we can find this object.
                 item.availableFrom.pop_front();
                 if (next.node != NULL)
                   {
-		    if (next.node->fDisconnect)  // Node was disconnected so we can't request from it
+                    // Do not request from this node if it was disconnected or the node pingtime is far beyond acceptable.
+                    // We only check pingtime during IBD because we don't want to lock vNodes too often and when the chain is syncd, waiting
+                    // just 5 seconds for a timeout is not an issue, however waiting for a slow node during IBD can really slow down the process.
+                    //   TODO: Eventually when we move away from vNodes or have a different mechanism for tracking ping times we can include
+                    //   this filtering in all our requests for blocks and transactions.
+		    if (next.node->fDisconnect || (!IsChainNearlySyncd() && !IsNodePingAcceptable(next.node)))
 		      {
-                      LOCK(cs_vNodes);
-                      LogPrint("req", "ReqMgr: %s removed ref to %d count %d (disconnect).\n",item.obj.ToString(), next.node->GetId(), next.node->GetRefCount());
-                      next.node->Release();
-                      next.node = NULL; // force the loop to get another node            
+                        LOCK(cs_vNodes);
+                        LogPrint("req", "ReqMgr: %s removed ref to %d count %d (disconnect).\n", item.obj.ToString(), next.node->GetId(), next.node->GetRefCount());
+                        next.node->Release();
+                        next.node = NULL; // force the loop to get another node            
 		      }
 		  }
 	        }
 
 	      if (next.node != NULL )
 		{
-		  if (item.lastRequestTime && IsChainNearlySyncd())  // if this is positive, we've requested at least once
+                  // If item.lastRequestTime is true then we've requested at least once and we'll try a re-request if the following conditions are met:
+                  //     The chain must be almost syncd and traffic shaping must not be turned on
+		  if (item.lastRequestTime && !IsTrafficShapingEnabled())
```

Comment:
My reasoning here was that if traffic shaping is on then and we lose a
block we'll still get it eventually when we hit the block download
timeout...as for txns it's no big deal IMO if we lose a few with traffic
shaping on.  It's obviously not a perfect solution but it does prevent a
lot of unecessary downloading of txns and blocks when the shaping is on,
which is exactly what the person who has shaping on is trying to
prevent.   That said, when doing IBD, the re-reqeusts will still happen
for blocks  to make sure the IBD process doesn't get hung up.

how can receivebufferrecentlyempty() help?


On 04/09/2016 3:45 AM, gandrewstone wrote:
>
> In src/requestManager.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79#discussion_r77450336>:
>
> >  		      }
> >  		  }
> >  	        }
> >  
> >  	      if (next.node != NULL )
> >  		{
> > -		  if (item.lastRequestTime && IsChainNearlySyncd())  // if this is positive, we've requested at least once
> > +                  // If item.lastRequestTime is true then we've requested at least once and we'll try a re-request if the following conditions are met:
> > +                  //     The chain must be almost syncd and traffic shaping must not be turned on
> > +		  if (item.lastRequestTime && !IsTrafficShapingEnabled())
>
> why no rerequest if traffic shaping? sure shaping could slow down the
> receipt of blocks, but clearly blocks could still be dropped if you
> are shaping or not. It seems like we have an interaction with traffic
> shaping that could be problematic -- that is, there exists
> pathological configurations where your own traffic shaping causes
> block rerequests. However turning off rerequesting if traffic shaping
> is turned on will cause runtime issues for drops during normal
> configuration.
>
> I think that the predicate we are looking for here is more like
> "ReceiveBufferRecentlyEmpty()"
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/79/files/efaaa2474cc017ebc1c84fa3748b50c293bff5c7#r77450336>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0MVnGdxYraRb5yway6IMOm0MCBx7ks5qmqE4gaJpZM4JokbP>.
>